pxr/usd/usdMtlx/reader.cpp

```diff
@@ -783,19 +784,7 @@ _NodeGraphBuilder::Build(ShaderNamesByOutputName* outputs)
         _AddNode(mtlxNode, usdPrim);
     }
     _ConnectNodes();
-
-    if (isInsideNodeGraph) {
-        _ConnectTerminals(_mtlxContainer, UsdShadeConnectableAPI(usdPrim));
-    }
-    else if (outputs) {
-        // Collect the outputs on the existing shader nodes.
-        for (mx::OutputPtr& mtlxOutput :
-                _mtlxContainer->getChildrenOfType<mx::Output>()) {
-            if (auto nodeName = _Attr(mtlxOutput, names.nodename)) {
-                (*outputs)[_Name(mtlxOutput)] = TfToken(nodeName);
-            }
-        }
-    }
```

Comment:
Could you explain why you deleted the second clause here?

Reply:
Firstly, credit to this goes to Dan Knowlton, who did the investigative work to figure this stuff out...

... So currently if there are "dangling" `<output>` elements at the top level of a document, then those outputs are never created, but the indirection implied by the `<output>` is resolved directly in the connections between the shader nodes.  

In the PR, the `"Nodegraphs"` prim, is now created at a `UsdShadeNodegraph` type, which then causes `UsdMtlx` to create the "dangling" `<output>` on the `"Nodegraphs"` prim itself, so we no longer need this map to track the indirection of the outputs.

Thats not the easiest explanation to grok, so here's a super small contrived example....

```
> cat example.mtlx
<?xml version="1.0"?>
 <materialx version="1.38">
   <ramplr name="layered_layer2_gradient" type="color3" />
   <output name="o_layered_layer2_gradient" type="color3" nodename="layered_layer2_gradient" />
   <nodedef name="ND_layerShader" node="layerShader">
     <input name="weight_1" type="color3" />
     <input name="weight_2" type="color3" />
     <input name="weight_3" type="color3" />
     <output name="out" type="surfaceshader" />
   </nodedef>
   <layerShader name="layered_sr" type="surfaceshader">
     <!-- An output coming from an "output" outside of an explicit node graph. -->
     <input name="weight_2" type="color3" output="o_layered_layer2_gradient" />
   </layerShader>
   <surfacematerial name="layered" type="material">
     <input name="surfaceshader" type="surfaceshader" nodename="layered_sr" />
   </surfacematerial>
 </materialx>
```

In USD 24.08 if we convert this to USD we get...
```
> usdcat example.mtlx
#usda 1.0

def "MaterialX"
{
    def "Materials"
    {
        def Material "layered"
        {
            color3f inputs:weight_1
            color3f inputs:weight_2
            color3f inputs:weight_3
            token outputs:mtlx:surface.connect = </MaterialX/Materials/layered/ND_layerShader.outputs:surface>

            def Shader "ND_layerShader" (
                prepend references = </MaterialX/Shaders/ND_layerShader>
            )
            {
                color3f inputs:weight_1.connect = </MaterialX/Materials/layered.inputs:weight_1>
                color3f inputs:weight_2.connect = </MaterialX/Materials/layered/NodeGraphs/layered_layer2_gradient.outputs:out>
                color3f inputs:weight_3.connect = </MaterialX/Materials/layered.inputs:weight_3>
            }

            def "NodeGraphs" (
                prepend references = </MaterialX/NodeGraphs>
            )
            {
            }
        }
    }

    def "Shaders"
    {
        def Shader "ND_layerShader"
        {
            uniform token info:id = "ND_layerShader"
            token outputs:surface
        }
    }

    def "NodeGraphs"
    {
        def Shader "layered_layer2_gradient"
        {
            uniform token info:id = "ND_ramplr_color3"
            color3f outputs:out
        }
    }
}
```

Notice here that there is no `output` in the `"Nodegraphs"` prim, but instead the `inputs:weight_2` on the `ND_layerShader` prim is directly connected to the `layered_layer2_gradient` prim inside the `"Nodegraphs"` prim.

With this patch applied we now get
```
> usdcat example.mtlx
#usda 1.0
(
    defaultPrim = "MaterialX"
    metersPerUnit = 0.01
    upAxis = "Y"
)

def "MaterialX"
{
    def "Materials"
    {
        def Material "layered"
        {
            color3f inputs:weight_1
            color3f inputs:weight_2
            color3f inputs:weight_3
            token outputs:mtlx:surface.connect = </MaterialX/Materials/layered/ND_layerShader.outputs:surface>

            def Shader "ND_layerShader" (
                prepend references = </MaterialX/Shaders/ND_layerShader>
            )
            {
                color3f inputs:weight_1.connect = </MaterialX/Materials/layered.inputs:weight_1>
                color3f inputs:weight_2.connect = </MaterialX/Materials/layered/NodeGraphs.outputs:o_layered_layer2_gradient>
                color3f inputs:weight_3.connect = </MaterialX/Materials/layered.inputs:weight_3>
            }

            def "NodeGraphs" (
                prepend references = </MaterialX/NodeGraphs>
            )
            {
            }
        }
    }

    def "Shaders"
    {
        def Shader "ND_layerShader"
        {
            uniform token info:id = "ND_layerShader"
            token outputs:surface
        }
    }

    def NodeGraph "NodeGraphs"
    {
        color3f outputs:o_layered_layer2_gradient.connect = </MaterialX/NodeGraphs/layered_layer2_gradient.outputs:out>

        def Shader "layered_layer2_gradient"
        {
            uniform token info:id = "ND_ramplr_color3"
            color3f outputs:out
        }
    }
}
```

Now with the change, that tags the `"Nodegraphs"` prim as a `UsdShadeNodegraph`, and thus allows connections through that prim, we add the `outputs:o_layered_layer2_gradient`, which itself is connected to the `layered_layer2_gradient` prim, and then the `inputs:weight_2` connection on `ND_layerShader`, is just connected to the new output on the `"Nodegraphs"` prim.

Because we're directly creating the same output and connection structure as is represented in the MaterialX document we no longer need the additional map to maintain the indirection.

Hopefully this makes sense - sorry for being so wordy :)

Reply:
Thanks for the great explanation!